File: python/taflow/indicators/fibonacci_fan.py
from typing import Any

import numpy as np

from .._native import FibonacciFan as _Native
from .._series import as_float64_series


class FibonacciFan:
    """Evaluate causal Fibonacci fan lines from a confirmed swing leg.

    Rust owns the persistent extrema, level arithmetic, warm-up, history, and
    processed-bar count. Required input order is high then low. The 3 output
    arrays are ordered as the 38.2%, 50.0%, and 61.8% fan-line prices. They are
    NaN until two alternating pivots establish the leg; later values extend
    those lines to the current bar without revising earlier history.
    append, extend, and reset are fluent. The external mapping is Wickra 0.9.9
    FibFan; TA-Lib has no equivalent chart-level function. TAFlow emits
    causal levels at each bar rather than a mutable chart annotation object.
    """

    def __init__(self, high: Any, low: Any) -> None:
        """Initialize and process required aligned chronological high/low data.

        Empty aligned series create fresh streaming state. Differing lengths
        raise ValueError before native state is mutated.
        """
        self._state = _Native()
        self.extend(high, low)

    def append(self, high: float, low: float) -> "FibonacciFan":
        """Append one high/low observation and return this adapter."""
        self._state.append(float(high), float(low))
        return self

    def extend(self, high: Any, low: Any) -> "FibonacciFan":
        """Append aligned high and low histories and return this adapter."""
        high_series = as_float64_series(high)
        low_series = as_float64_series(low)
        if len(high_series) != len(low_series):
            raise ValueError("high and low inputs must have equal lengths")
        self._state.extend(high_series, low_series)
        return self

    @property
    def value(self) -> tuple[float, float, float] | None:
        """Return the latest ordered native levels, or None during warm-up."""
        return self._state.value

    def compute(self) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
        """Return the 3 aligned native output histories in documented order."""
        return self._state.compute()

    def reset(self) -> "FibonacciFan":
        """Restore fresh native state without reallocating and return self."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the processed-bar count delegated to native state."""
        return len(self._state)

File: crates/taflow-core/src/indicators/fibonacci_fan.rs
use super::pattern_swing::{SwingTracker, SWING_THRESHOLD};
use crate::error::TaResult;

/// Fibonacci fan-line prices at the current bar.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct FibonacciFanValue {
    pub fan_382: f64,
    pub fan_500: f64,
    pub fan_618: f64,
}

/// Fan lines from the latest confirmed swing leg.
#[derive(Debug, Clone)]
pub struct FibonacciFan {
    swing: SwingTracker,
    count: usize,
    value: Option<FibonacciFanValue>,
}

impl FibonacciFan {
    /// Create an empty fan tracker.
    pub fn new() -> TaResult<Self> {
        Ok(Self {
            swing: SwingTracker::new(SWING_THRESHOLD, 2),
            count: 0,
            value: None,
        })
    }

    /// Append one high/low bar and evaluate three fan lines.
    pub fn append(&mut self, high: f64, low: f64) -> Option<FibonacciFanValue> {
        self.count += 1;
        self.swing.append(high, low);
        let pivots = self.swing.pivots();
        self.value = if pivots.len() < 2 {
            None
        } else {
            let start = pivots[0];
            let end = pivots[1];
            let progress =
                (self.swing.current_bar() - start.bar) as f64 / (end.bar - start.bar) as f64;
            let line = |ratio: f64| start.price + ratio * (end.price - start.price) * progress;
            Some(FibonacciFanValue {
                fan_382: line(0.382),
                fan_500: line(0.500),
                fan_618: line(0.618),
            })
        };
        self.value
    }

    /// Return the latest fan-line prices.
    pub fn value(&self) -> Option<FibonacciFanValue> {
        self.value
    }
    /// Return the processed-bar count.
    pub fn len(&self) -> usize {
        self.count
    }
    /// Return whether no bars were processed.
    pub fn is_empty(&self) -> bool {
        self.count == 0
    }
    /// Clear pivots and output.
    pub fn reset(&mut self) {
        self.swing.reset();
        self.count = 0;
        self.value = None;
    }
}
